provider/sma/discover.go

```diff
@@ -0,0 +1,163 @@
+
+// map of created discover instances
+var discoverers = make(map[string]*Discoverer)
+var discoverersMutex sync.Mutex
+
+// initialize sunny logger only once
+var once sync.Once
+
+// GetDiscoverer fo the given interface
+func GetDiscoverer(iface string) (*Discoverer, error) {
+	// on time initialization of sunny logger
+	log := util.NewLogger("sma")
+	once.Do(func() {
+		sunny.Log = log.TRACE
+	})
+
+	discoverersMutex.Lock()
+	defer discoverersMutex.Unlock()
+
+	// get or create discoverer
+	discoverer, ok := discoverers[iface]
+	if !ok {
+		conn, err := sunny.NewConnection(iface)
+		if err != nil {
+			return nil, fmt.Errorf("connection failed: %w", err)
+		}
+
+		discoverer = &Discoverer{
+			log:     log,
+			conn:    conn,
+			devices: make(map[uint32]*Device),
+		}
+
+		go discoverer.run()
+
+		discoverers[iface] = discoverer
+	}
+	return discoverer, nil
+}
+
+// Discoverer discovers SMA devicesBySerial in background while providing already found devicesBySerial
+type Discoverer struct {
+	log     *util.Logger
+	conn    *sunny.Connection
+	devices map[uint32]*Device
+	mux     sync.RWMutex
+	done    uint32
+}
+
+func (d *Discoverer) createDevice(device *sunny.Device) *Device {
+	dev := &Device{
+		Device: device,
+		log:    d.log,
+		mux:    util.NewWaiter(udpTimeout, func() { d.log.TRACE.Println("wait for initial value") }),
+		values: make(map[sunny.ValueID]interface{}),
+	}
+
+	go func() {
+		for range time.NewTicker(time.Second).C {
+			err := dev.updateValues()
+			if err != nil {
+				d.log.ERROR.Println(err)
+			}
+		}
+	}()
+
+	return dev
+}
+
+func (d *Discoverer) addDevice(device *sunny.Device) {
+	d.mux.Lock()
+	defer d.mux.Unlock()
+
+	if _, ok := d.devices[device.SerialNumber()]; !ok {
+		d.devices[device.SerialNumber()] = d.createDevice(device)
+	}
+}
+
+// run discover and store found devicesBySerial
+func (d *Discoverer) run() {
+	devices := make(chan *sunny.Device)
+
+	go func() {
+		for device := range devices {
+			d.addDevice(device)
+		}
+	}()
+
+	// discover devicesBySerial and wait for results
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	d.conn.DiscoverDevices(ctx, devices, "")
+	cancel()
+	close(devices)
+
+	// mark discover as done
+	atomic.AddUint32(&d.done, 1)
+}
+
+func (d *Discoverer) get(serial uint32, password string) *Device {
+	d.mux.RLock()
+	defer d.mux.RUnlock()
+
+	device := d.devices[serial]
+	if device != nil {
+		device.SetPassword(password)
+	}
+	return device
+}
+
+// DeviceBySerial with the given serial number
+func (d *Discoverer) DeviceBySerial(serial uint32, password string) *Device {
+	start := time.Now()
+	for time.Since(start) < time.Second*3 {
+		// discover done -> return immediately regardless of result
+		if atomic.LoadUint32(&d.done) != 0 {
+			return d.get(serial, password)
+		}
+
+		// Device with serial found -> return
+		if device := d.get(serial, password); device != nil {
+			return device
+		}
+
+		time.Sleep(time.Millisecond * 10)
```

Comment:
Es wäre schön dieses aktive warten z.B. durch einen Channel zu ersetzen aus dem bis Timeout gelesen wird:

	select{
	case <-time.NewTimer(3*time.Second):
		return false
	case device:=<-found:
		return device
	}

Reply:
wenn du mir sagst wie ich mit einem channel einen lookup in einer list mache und gleichzeitig auf neue Geräte warten kann.
Ich hab da bisher keine bessere Lösung gefunden

Reply:
Das ganze Teil hier:

	go func() {
		for device := range devices {
			d.addDevice(device)
		}
	}()

müsste so gestaltet sein, dass Subscriber sich dran hängen können und auch gleich eine Nachricht bekommen. Im Prinzip ein Slice of Channels die alle das Device zugeschoben bekommen. Ich vermute es wäre aber mehr Code, also lassen wir das gerne :O. Ist einfach nciht ganz "schön".

Reply:
nicht nur das. du bekommst auch noch das Problem wenn der Subscriber sich erst dann meldet wenn die ersten x Geräte schon gefunden wurden.

Man müsste also atomic nen lookup in der liste der bereits gefunden Geräte machen und dann nen channel returnenn für später gefundene. Das ist bestimmt möglich aber ich bin mir aktuell nicht mal so ganz sicher wie man sowas am besten umsetzt.